packages/dds/matrix/src/matrix.ts

```diff
@@ -264,20 +325,27 @@ export class SharedMatrix<T = any>
 		rowHandle = this.rows.getAllocatedHandle(row),
 		colHandle = this.cols.getAllocatedHandle(col),
 	) {
-		if (this.undo !== undefined) {
-			let oldValue = this.cells.getCell(rowHandle, colHandle);
-			if (oldValue === null) {
-				oldValue = undefined;
+		this.protectAgainstReentrancy(() => {
+			if (this.undo !== undefined) {
+				let oldValue = this.cells.getCell(rowHandle, colHandle);
+				if (oldValue === null) {
+					oldValue = undefined;
+				}
+
+				this.undo.cellSet(rowHandle, colHandle, oldValue);
 			}
 
-			this.undo.cellSet(rowHandle, colHandle, oldValue);
-		}
+			this.cells.setCell(rowHandle, colHandle, value);
```

Comment:
I wonder if this is correct, and what is the correct way to handle it.
Do we undo this operation if it happens to be failed FWW operation?

Reply:
Here is where I have trouble seeing how to do it correctly - one of the cases to take care of:
Let's imagine we are in LWW mode. Client would need to make local changes. But it might be disconnected, and by time it reconnects, it can find itself in FWW mode.
So, I think applying local changes for all modes is correct, but this means that we have to be able to undo this change if it results in a conflict later on.

I'd recommend first to cover all these cases in UTs, and ensure they are failing, before coming up with solutions.

Reply:
Ah, maybe it's not that bad, if we failed, we should already processed some op that overwritten our value in matrix, right? I still worry about all the corner cases - let's make sure we cover them in UT, and in fuzz tests

Reply:
Will handle it separately. This seems to be problem in LWW also.